File: src/utils/storage.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

import { Script, IdeaNote } from '../types';

const SAMPLE_SCRIPT: Script = {
  id: 'sample_cyber_dawn',
  title: 'Cyber Dawn',
  writer: 'Cassandra Sterling',
  email: 'cassandra@sterlingfx.com',
  phone: '+1 (555) 019-2831',
  address: '102 Pine Street\nLos Angeles, CA 90028',
  notes: 'First draft. A speculative sci-fi thriller.',
  createdAt: new Date().toISOString(),
  updatedAt: new Date().toISOString(),
  content: [
    { id: '1', format: 'scene-heading', text: 'INT. STERLING LABS - NIGHT' },
    { id: '2', format: 'action', text: 'Flashing amber server lights paint the concrete bunker in deep crimson. A low hum vibrates the glass partition.' },
    { id: '3', format: 'character', text: 'LEO' },
    { id: '4', format: 'parenthetical', text: 'staring at a cascade of green code' },
    { id: '5', format: 'dialogue', text: 'It\'s repeating. Over and over again. Every line is an invitation.' },
    { id: '6', format: 'character', text: 'SARA' },
    { id: '7', format: 'dialogue', text: 'Don\'t run the compiler. We don\'t know how it behaves in isolated sandboxes.' },
    { id: '8', format: 'character', text: 'LEO' },
    { id: '9', format: 'parenthetical', text: 'smiles, hand hovering over key' },
    { id: '10', format: 'dialogue', text: 'That\'s the beauty of it. It doesn\'t want to be isolated.' },
    { id: '11', format: 'transition', text: 'CUT TO:' },
    { id: '12', format: 'scene-heading', text: 'EXT. LOS ANGELES STREET - DEEPER NIGHT' },
    { id: '13', format: 'action', text: 'Neon billboards flicker against the overcast sky. Rain begins to pepper the black asphalt.' },
    { id: '14', format: 'shot', text: 'WIDE ANCHOR SHOT - THE MAIN DATACENTER' },
    { id: '15', format: 'action', text: 'A single backup generator begins to sputter as all the city lights below start to slowly wind down in sequence.' }
  ]
};

export const Storage = {
  getScripts(): Script[] {
    try {
      const stored = localStorage.getItem('screenwriter_scripts');
      if (!stored) {
        // Initialize with sample script
        const scripts = [SAMPLE_SCRIPT];
        localStorage.setItem('screenwriter_scripts', JSON.stringify(scripts));
        return scripts;
      }
      return JSON.parse(stored);
    } catch {
      return [SAMPLE_SCRIPT];
    }
  },

  getScript(id: string): Script | null {
    const scripts = this.getScripts();
    return scripts.find((s) => s.id === id) || null;
  },

  saveScript(script: Script): void {
    const scripts = this.getScripts();
    const idx = scripts.findIndex((s) => s.id === script.id);
    const updatedScript = {
      ...script,
      updatedAt: new Date().toISOString()
    };
    if (idx >= 0) {
      scripts[idx] = updatedScript;
    } else {
      scripts.unshift(updatedScript);
    }
    localStorage.setItem('screenwriter_scripts', JSON.stringify(scripts));
  },

  deleteScript(id: string): void {
    const scripts = this.getScripts().filter((s) => s.id !== id);
    localStorage.setItem('screenwriter_scripts', JSON.stringify(scripts));
  },

  getNotes(): IdeaNote[] {
    try {
      const stored = localStorage.getItem('screenwriter_notes');
      if (!stored) {
        const sampleNote: IdeaNote = {
          id: 'sample_idea_noir',
          title: 'Sci-Fi Setting Brainstorm',
          description: 'Initial worldbuilding parameters and character arcs.',
          content: '<h2><strong>The World of Cyber Dawn</strong></h2><p>Here are the core atmospheric parameters for the world of Cyber Dawn:</p><ul><li><strong>Setting:</strong> Outpost Omega, a neon-lit rain-slicked metropolis covered by an artificial climate dome.</li><li><strong>Visual Theme:</strong> Retro-futuristic brutalist towers paired with vibrant analog cathode screens and copper wiring.</li><li><strong>Leo\'s Motivation:</strong> Leo believes the code is sentient. His obsession is to decrypt the repeating block.</li><li><strong>Sara\'s Conflict:</strong> Sara is haunted by a previous sandbox incident that crashed an entire sector\'s grid.</li></ul>',
          createdAt: new Date().toISOString(),
          updatedAt: new Date().toISOString()
        };
        const notes = [sampleNote];
        localStorage.setItem('screenwriter_notes', JSON.stringify(notes));
        return notes;
      }
      return JSON.parse(stored);
    } catch {
      return [];
    }
  },

  getNote(id: string): IdeaNote | null {
    const notes = this.getNotes();
    return notes.find((n) => n.id === id) || null;
  },

  saveNote(note: IdeaNote): void {
    const notes = this.getNotes();
    const idx = notes.findIndex((n) => n.id === note.id);
    const updatedNote = {
      ...note,
      updatedAt: new Date().toISOString()
    };
    if (idx >= 0) {
      notes[idx] = updatedNote;
    } else {
      notes.unshift(updatedNote);
    }
    localStorage.setItem('screenwriter_notes', JSON.stringify(notes));
  },

  deleteNote(id: string): void {
    const notes = this.getNotes().filter((n) => n.id !== id);
    localStorage.setItem('screenwriter_notes', JSON.stringify(notes));
  }
};

File: app/src/main/java/com/simbi/screenwriter/Storage.kt
package com.simbi.screenwriter

import android.content.Context
import android.content.SharedPreferences
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

object Storage {
    private const val PREFS_NAME = "screenwriter_prefs"
    private const val KEY_SCRIPTS = "screenwriter_scripts"
    private const val KEY_NOTES = "screenwriter_notes"

    private val gson = Gson()

    private fun getISO8601String(): String {
        val tz = TimeZone.getTimeZone("UTC")
        val df = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US)
        df.timeZone = tz
        return df.format(Date())
    }

    private fun getPrefs(context: Context): SharedPreferences {
        return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    private val SAMPLE_SCRIPT = Script(
        id = "sample_cyber_dawn",
        title = "Cyber Dawn",
        writer = "Cassandra Sterling",
        email = "cassandra@sterlingfx.com",
        phone = "+1 (555) 019-2831",
        address = "102 Pine Street\nLos Angeles, CA 90028",
        notes = "First draft. A speculative sci-fi thriller.",
        createdAt = "2026-06-11T09:41:00.000Z",
        updatedAt = "2026-06-11T09:41:00.000Z",
        content = listOf(
            ScreenplayLine("1", "scene-heading", "INT. STERLING LABS - NIGHT"),
            ScreenplayLine("2", "action", "Flashing amber server lights paint the concrete bunker in deep crimson. A low hum vibrates the glass partition."),
            ScreenplayLine("3", "character", "LEO"),
            ScreenplayLine("4", "parenthetical", "staring at a cascade of green code"),
            ScreenplayLine("5", "dialogue", "It's repeating. Over and over again. Every line is an invitation."),
            ScreenplayLine("6", "character", "SARA"),
            ScreenplayLine("7", "dialogue", "Don't run the compiler. We don't know how it behaves in isolated sandboxes."),
            ScreenplayLine("8", "character", "LEO"),
            ScreenplayLine("9", "parenthetical", "smiles, hand hovering over key"),
            ScreenplayLine("10", "dialogue", "That's the beauty of it. It doesn't want to be isolated."),
            ScreenplayLine("11", "transition", "CUT TO:"),
            ScreenplayLine("12", "scene-heading", "EXT. LOS ANGELES STREET - DEEPER NIGHT"),
            ScreenplayLine("13", "action", "Neon billboards flicker against the overcast sky. Rain begins to pepper the black asphalt."),
            ScreenplayLine("14", "shot", "WIDE ANCHOR SHOT - THE MAIN DATACENTER"),
            ScreenplayLine("15", "action", "A single backup generator begins to sputter as all the city lights below start to slowly wind down in sequence.")
        )
    )

    private val SAMPLE_NOTE = IdeaNote(
        id = "sample_idea_noir",
        title = "Sci-Fi Setting Brainstorm",
        description = "Initial worldbuilding parameters and character arcs.",
        content = "<h2><strong>The World of Cyber Dawn</strong></h2><p>Here are the core atmospheric parameters for the world of Cyber Dawn:</p><ul><li><strong>Setting:</strong> Outpost Omega, a neon-lit rain-slicked metropolis covered by an artificial climate dome.</li><li><strong>Visual Theme:</strong> Retro-futuristic brutalist towers paired with vibrant analog cathode screens and copper wiring.</li><li><strong>Leo's Motivation:</strong> Leo believes the code is sentient. His obsession is to decrypt the repeating block.</li><li><strong>Sara's Conflict:</strong> Sara is haunted by a previous sandbox incident that crashed an entire sector's grid.</li></ul>",
        createdAt = "2026-06-11T09:41:00.000Z",
        updatedAt = "2026-06-11T09:41:00.000Z"
    )

    fun getScripts(context: Context): List<Script> {
        val prefs = getPrefs(context)
        val json = prefs.getString(KEY_SCRIPTS, null)
        if (json == null) {
            val scripts = listOf(SAMPLE_SCRIPT)
            saveScriptsList(context, scripts)
            return scripts
        }
        return try {
            val type = object : TypeToken<List<Script>>() {}.type
            gson.fromJson(json, type) ?: listOf(SAMPLE_SCRIPT)
        } catch (e: Exception) {
            listOf(SAMPLE_SCRIPT)
        }
    }

    private fun saveScriptsList(context: Context, scripts: List<Script>) {
        getPrefs(context).edit().putString(KEY_SCRIPTS, gson.toJson(scripts)).apply()
    }

    fun getScript(context: Context, id: String): Script? {
        return getScripts(context).find { it.id == id }
    }

    fun saveScript(context: Context, script: Script) {
        val scripts = getScripts(context).toMutableList()
        val idx = scripts.indexOfFirst { it.id == script.id }
        script.updatedAt = getISO8601String()
        if (idx >= 0) {
            scripts[idx] = script
        } else {
            scripts.add(0, script)
        }
        saveScriptsList(context, scripts)
    }

    fun deleteScript(context: Context, id: String) {
        val scripts = getScripts(context).filter { it.id != id }
        saveScriptsList(context, scripts)
    }

    fun getNotes(context: Context): List<IdeaNote> {
        val prefs = getPrefs(context)
        val json = prefs.getString(KEY_NOTES, null)
        if (json == null) {
            val notes = listOf(SAMPLE_NOTE)
            saveNotesList(context, notes)
            return notes
        }
        return try {
            val type = object : TypeToken<List<IdeaNote>>() {}.type
            gson.fromJson(json, type) ?: listOf(SAMPLE_NOTE)
        } catch (e: Exception) {
            listOf(SAMPLE_NOTE)
        }
    }

    private fun saveNotesList(context: Context, notes: List<IdeaNote>) {
        getPrefs(context).edit().putString(KEY_NOTES, gson.toJson(notes)).apply()
    }

    fun getNote(context: Context, id: String): IdeaNote? {
        return getNotes(context).find { it.id == id }
    }

    fun saveNote(context: Context, note: IdeaNote) {
        val notes = getNotes(context).toMutableList()
        val idx = notes.indexOfFirst { it.id == note.id }
        note.updatedAt = getISO8601String()
        if (idx >= 0) {
            notes[idx] = note
        } else {
            notes.add(0, note)
        }
        saveNotesList(context, notes)
    }

    fun deleteNote(context: Context, id: String) {
        val notes = getNotes(context).filter { it.id != id }
        saveNotesList(context, notes)
    }
}
